### PR TITLE
KEYCLOAK-12175 - Platform SPI

### PR DESCRIPTION
**Pull request by dteleguin**
Initial implementation of the Platform SPI

### CODE REVIEW

File: distribution/feature-packs/server-feature-pack/src/main/resources/modules/system/layers/keycloak/org/keycloak/keycloak-server-spi-private/main/module.xml [hunk 1]
```diff
@@ -30,6 +30,7 @@
         <module name="org.keycloak.keycloak-common"/>
         <module name="org.keycloak.keycloak-core"/>
         <module name="org.keycloak.keycloak-server-spi"/>
+        <module name="org.keycloak.keycloak-wildfly-extensions" services="import"/>
```
Comment (stianst) [inline distribution/feature-packs/server-feature-pack/src/main/resources/modules/system/layers/keycloak/org/keycloak/keycloak-server-spi-private/main/module.xml]: Shouldn't be needed as Platform SPI should be moved to services

File: distribution/feature-packs/server-feature-pack/src/main/resources/modules/system/layers/keycloak/org/keycloak/keycloak-server-subsystem/dependencies/main/module.xml [hunk 2]
```diff
@@ -23,6 +23,7 @@
 
     <dependencies>
         <module name="org.keycloak.keycloak-services" export="true"/>
+        <module name="org.keycloak.keycloak-wildfly-extensions" export="true"/>
```
Comment (stianst) [inline distribution/feature-packs/server-feature-pack/src/main/resources/modules/system/layers/keycloak/org/keycloak/keycloak-server-subsystem/dependencies/main/module.xml]: This shouldn't be needed as platform provider is used within services only, which has a dependency on keycloak-wildfly-extensions

File: quarkus/extensions/pom.xml [hunk 3]
```diff
@@ -32,6 +32,14 @@
                 </exclusion>
             </exclusions>
         </dependency>
+        <dependency>
```
Comment (stianst) [inline quarkus/extensions/pom.xml]: I can't see commons-lang being used anywhere. Is it needed?

File: server-spi-private/src/main/java/org/keycloak/platform/Platform.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
+/*
```
Comment (stianst) [inline server-spi-private/src/main/java/org/keycloak/platform/Platform.java]: Should be moved to services

File: server-spi-private/src/main/java/org/keycloak/platform/PlatformProvider.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+package org.keycloak.platform;
+
+public interface PlatformProvider {
+    
+    void onStartup(Runnable runnable);
+
+    void onShutdown(Runnable runnable);
+
+    void die(Throwable cause);
```
Comment (stianst) [inline server-spi-private/src/main/java/org/keycloak/platform/PlatformProvider.java]: Perhaps "exit(Throwable cause)" would be a better name?

File: distribution/feature-packs/server-feature-pack/src/main/resources/modules/system/layers/keycloak/org/keycloak/keycloak-server-spi-private/main/module.xml (hunk 1 above)
Comment (dteleguin) [reply distribution/feature-packs/server-feature-pack/src/main/resources/modules/system/layers/keycloak/org/keycloak/keycloak-server-spi-private/main/module.xml]: done

File: distribution/feature-packs/server-feature-pack/src/main/resources/modules/system/layers/keycloak/org/keycloak/keycloak-server-subsystem/dependencies/main/module.xml (hunk 2 above)
Comment (dteleguin) [reply distribution/feature-packs/server-feature-pack/src/main/resources/modules/system/layers/keycloak/org/keycloak/keycloak-server-subsystem/dependencies/main/module.xml]: done; had to add `export="true"` to `services` → `wildfy-extensions` dep to make the latter visible to the WAR

File: quarkus/extensions/pom.xml (hunk 3 above)
Comment (dteleguin) [reply quarkus/extensions/pom.xml]: sorry, sneaked in by mistake

File: server-spi-private/src/main/java/org/keycloak/platform/Platform.java (hunk 4 above)
Comment (dteleguin) [reply server-spi-private/src/main/java/org/keycloak/platform/Platform.java]: done

File: server-spi-private/src/main/java/org/keycloak/platform/PlatformProvider.java (hunk 5 above)
Comment (dteleguin) [reply server-spi-private/src/main/java/org/keycloak/platform/PlatformProvider.java]: done

**Review by pedroigor — DISMISSED** [review]

File: quarkus/extensions/pom.xml [hunk 6]
```diff
@@ -32,6 +32,10 @@
...
```
Comment (stianst) [inline quarkus/extensions/pom.xml]: Is this needed now that the Provider is moved to services?

File: services/src/main/java/org/keycloak/services/resources/KeycloakApplication.java [hunk 7]
```diff
@@ -97,14 +95,11 @@ public KeycloakApplication() {
 
         try {
 
+            logger.infov("Using PlatformProvider: {0}", platform.getClass().getName());
```
Comment (stianst) [inline services/src/main/java/org/keycloak/services/resources/KeycloakApplication.java]: Should be debug level

**Review by stianst — APPROVED** [review]

**Review by stianst — CHANGES_REQUESTED** [review]
quarkus/extensions/pom.xml probably doesn't need a dep on server-spi-private. 

Log platform provider in KeycloakApplication should be debug level, not info

File: services/src/main/java/org/keycloak/services/resources/KeycloakApplication.java (hunk 7 above)
Comment (dteleguin) [reply services/src/main/java/org/keycloak/services/resources/KeycloakApplication.java]: done; fixed for ConfigProvider as well

File: quarkus/extensions/pom.xml (hunk 6 above)
Comment (dteleguin) [reply quarkus/extensions/pom.xml]: done

**Review by stianst — APPROVED** [review]

**Pull request merged**